File: KahunaJailBreakDetection/Classes/KahunaJailBreakDetection.m
//
//  KahunaJailBreakDetection.m
//  KahunaJailBreakDetection
//
//  Created by Siddharth on 7/210/17.
//  Copyright © 2017 Kahuna Systems. All rights reserved.
//

#import "KahunaJailBreakDetection.h"
//#import "KahunaJailBreakViewController.h"

@interface KahunaJailBreakDetection ()

@property(strong,nonatomic) UIViewController *jailBreakViewController;

@end

@implementation KahunaJailBreakDetection

+(id)sharedInstance {
    static KahunaJailBreakDetection *__sharedInstance;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        __sharedInstance = [[KahunaJailBreakDetection alloc]init];
    });
    return __sharedInstance;
}

-(void)setYourViewController:(UIViewController *)viewController{
    self.jailBreakViewController = viewController;
}

+ (BOOL)isJailbroken
{
#if !(TARGET_IPHONE_SIMULATOR)

    FILE *file = fopen("/Applications/Cydia.app", "r");
    if (file) {
        fclose(file);
        return YES;
    }
    file = fopen("/Library/MobileSubstrate/MobileSubstrate.dylib", "r");
    if (file) {
        fclose(file);
        return YES;
    }
    file = fopen("/bin/bash", "r");
    if (file) {
        fclose(file);
        return YES;
    }
    file = fopen("/usr/sbin/sshd", "r");
    if (file) {
        fclose(file);
        return YES;
    }
    file = fopen("/etc/apt", "r");
    if (file) {
        fclose(file);
        return YES;
    }
    file = fopen("/usr/bin/ssh", "r");
    if (file) {
        fclose(file);
        return YES;
    }
    
    NSFileManager *fileManager = [NSFileManager defaultManager];

    if ([fileManager fileExistsAtPath:@"/Applications/Cydia.app"]) {
        return YES;
    } else if ([fileManager fileExistsAtPath:@"/Library/MobileSubstrate/MobileSubstrate.dylib"]) {
        return YES;
    } else if ([fileManager fileExistsAtPath:@"/bin/bash"]) {
        return YES;
    } else if ([fileManager fileExistsAtPath:@"/usr/sbin/sshd"]) {
        return YES;
    } else if ([fileManager fileExistsAtPath:@"/etc/apt"]) {
        return YES;
    } else if ([fileManager fileExistsAtPath:@"/usr/bin/ssh"]) {
        return YES;
    }
    
    // Check if the app can access outside of its sandbox
    NSError *error = nil;
    NSString *string = @".";
    [string writeToFile:@"/private/jailbreak.txt" atomically:YES encoding:NSUTF8StringEncoding error:&error];
    if (!error) {
        return YES;
    } else {
        [fileManager removeItemAtPath:@"/private/jailbreak.txt" error:nil];
    }
    // Check if the app can open a Cydia's URL scheme
    if ([[UIApplication sharedApplication] canOpenURL:[NSURL URLWithString:@"cydia://package/com.example.package"]]) {
        return YES;
    }
#endif
    return NO;
}

-(void)checkJailDeviceinDevice{
    if ([KahunaJailBreakDetection isJailbroken]) {
        NSString *appName = [[[NSBundle mainBundle]infoDictionary]objectForKey:@"CFBundleName"];
        NSString *appDesc = @"";
        if(appName.length > 0){
            appDesc = [NSString stringWithFormat:@"It appears that your device is compromised or rooted and %@ can not continue. Please contact our support team for further guidance.",appName];
        }
        NSDictionary *attrDict = @{
                                   NSFontAttributeName : [UIFont boldSystemFontOfSize:21.0],
                                   NSForegroundColorAttributeName : [UIColor redColor]
                                   };
        NSMutableAttributedString *attributedString = [[NSMutableAttributedString alloc] initWithString:@"Security Error" attributes:attrDict];
        NSDictionary *attrDict1 = @{
                                    NSForegroundColorAttributeName : [UIColor blackColor]
                                    };
        NSMutableAttributedString *attributedString1 = [[NSMutableAttributedString alloc] initWithString:appDesc attributes:attrDict1];
        
        NSRange boldRange = [appDesc rangeOfString:appName];
        [attributedString1 addAttribute: NSFontAttributeName value:[UIFont boldSystemFontOfSize:17] range:boldRange];
        UIAlertController *alertController=[UIAlertController alertControllerWithTitle:@""
                                                                               message:@""
                                                                        preferredStyle:UIAlertControllerStyleAlert];
        [alertController setValue:attributedString forKey:@"attributedTitle"];
        [alertController setValue:attributedString1 forKey:@"attributedMessage"];
        if(self.jailBreakViewController!=nil){
            [self.jailBreakViewController presentViewController:alertController animated:true completion:nil];
        }
    }
}

@end

File: KahunaJailBreakDetection/Classes/KahunaJailBreakDetection.swift
//
//  KahunaJailBreakDetection.swift
//  KahunaJailBreakDetection
//
//  Created by Siddharth Chopra on 21/02/18.
//

import Foundation
import UIKit

public class KahunaJailBreakDetection: NSObject {

    public static let sharedInstance = KahunaJailBreakDetection()
    var jailBreakViewController: UIViewController?

    public override init() {
        super.init()
    }

    public func setYourViewController(_ viewController: UIViewController) {
        self.jailBreakViewController = viewController
    }

    public class func isJailbroken() -> Bool {
        if TARGET_IPHONE_SIMULATOR != 1 {
            // Check existence of files that are common for jailbroken devices
            if FileManager.default.fileExists(atPath: "/Applications/Cydia.app")
                || FileManager.default.fileExists(atPath: "/Library/MobileSubstrate/MobileSubstrate.dylib")
                || FileManager.default.fileExists(atPath: "/bin/bash")
                || FileManager.default.fileExists(atPath: "/usr/sbin/sshd")
                || FileManager.default.fileExists(atPath: "/etc/apt")
                || FileManager.default.fileExists(atPath: "/private/var/lib/apt/")
                || FileManager.default.fileExists(atPath: "/usr/bin/ssh") {
                return true
            }
            // Check if the app can open a Cydia's URL scheme
            if let aString = URL(string: "cydia://package/com.example.package") {
                if UIApplication.shared.canOpenURL(aString) {
                    return true
                }
            }
            // Check if the app can access outside of its sandbox
            let stringToWrite = "Jailbreak Test"
            do {
                try stringToWrite.write(toFile: "/private/jailbreak.txt", atomically: true, encoding: .utf8)
                //Device is jailbroken
                return true
            } catch {
                return false
            }
        } else {
            return false
        }
    }

    public func checkJailDeviceinDevice() {
        if KahunaJailBreakDetection.isJailbroken() {
            var appDesc = ""
            var appName = ""
            if let appname = Bundle.main.infoDictionary?["CFBundleName"] as? String, appname.count > 0 {
                appDesc = "It appears that your device is compromised or rooted and \(appname) can not continue. Please contact our support team for further guidance."
                appName = appname
            }
            let attrDict = [NSFontAttributeName: UIFont.boldSystemFont(ofSize: 21.0), NSForegroundColorAttributeName: UIColor.red]
            let attributedString = NSMutableAttributedString(string: "Security Error", attributes: attrDict)
            let attrDict1 = [NSForegroundColorAttributeName: UIColor.black]
            let attributedString1 = NSMutableAttributedString(string: appDesc, attributes: attrDict1)
            let boldRange: NSRange? = (appDesc as NSString).range(of: appName)
            if let aRange = boldRange {
                attributedString1.addAttribute(NSFontAttributeName, value: UIFont.boldSystemFont(ofSize: 17), range: aRange)
            }
            let alertController = UIAlertController(title: "", message: "", preferredStyle: .alert)
            alertController.setValue(attributedString, forKey: "attributedTitle")
            alertController.setValue(attributedString1, forKey: "attributedMessage")
            if self.jailBreakViewController != nil {
                self.jailBreakViewController?.present(alertController, animated: true) { () -> Void in }
            }
        }
    }

}

